# From a MNIST handwritten digit (SVM) classifier to Apple's Core ML format

This tutorial will NOT go in depth of machine learning topics, it is a short tutorial on how to convert scikit-learn SVM model Apple's coreml format. 

In [30]:
# import mnist_loader_py2 as mnist_loader  # enable this if python 2 is used
import mnist_loader_py3 as mnist_loader
from sklearn import svm
import platform
print(platform.python_version())
import sys
print(sys.executable)

3.6.6
/Users/erkandiken/Development/deep-learning/mnist_classifier/mnist-scikit-svm-to-coreml/bin/python


# 1.  Loading the MNIST image data

First, we will load the MNIST image data. Here, we do this with the help of an external library (mnist_loader).

You can find the original python file in the following github repo: https://github.com/mnielsen/neural-networks-and-deep-learning/tree/master/src

In [31]:
training_data, validation_data, test_data = mnist_loader.load_data()

We will not be using the validation data, however, we kept the standard way of data partitining.

Training data is returned as a tuple of two entires. The first entry contains the actual training images in a vectorized form (two dimensional 28 pixels by 28 pixels image is flattend to 1 x 784 sized vector). In total, we have 50K traning images, hence, 50K x 784 sized training data. The second entry contains the true label (in this case a digit between 0 and 9) of each training image. 

In [32]:
print (type(training_data))
print ("First entry: ", training_data[0].shape)
print ("Second entry: ", training_data[1].shape)
print ("Indeed, each entry of training data is numpy ndarray:")
print (type(training_data[0]))

<class 'tuple'>
First entry:  (50000, 784)
Second entry:  (50000,)
Indeed, each entry of training data is numpy ndarray:
<class 'numpy.ndarray'>


In [33]:
print ("true label of the first image is : ", training_data[1][0])

true label of the first image is :  5


Test data has the same structure, however, only with 10K images.

In [34]:
print (type(test_data))
print ("First entry: ", test_data[0].shape)
print ("Second entry: ", test_data[1].shape)
print (type(test_data[0]))

<class 'tuple'>
First entry:  (10000, 784)
Second entry:  (10000,)
<class 'numpy.ndarray'>


# 2. Train the model

Before going to model training, I will rename the training and test data components.

In [35]:
X_train = training_data[0]
y_train = training_data[1]
X_test = test_data[0]
y_test = test_data[1]

We use the Support Vector Machines (SVM) from scikit-learn library to build our ML model. I will use only first 1000 data images for fitting in order to quickly check if the model is built successfully.

In [36]:
X_train_sub = X_train[:1000]
y_train_sub = y_train[:1000]
svm_model = svm.SVC(gamma='auto')
svm_model.fit(X_train_sub, y_train_sub)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

SVM model is created. We can immediately check what will be the prediction of the model for the first digit in the test data.

In [37]:
predicted_test_0 = svm_model.predict(X_test[0].reshape(1, -1))
print("predicted: ", int(predicted_test_0), "true value: ", y_test[0])

predicted:  7 true value:  7


You should see 7 for the predicted and true label of the digit. Even if we used 1K data out of 50K data for training, our model was good enough to predict the first digit.

We can check the performance of the model on whole test data.

In [38]:
# testing
predictions = [int(a) for a in svm_model.predict(X_test)]
num_correct = sum(int(a == y) for a, y in zip(predictions, y_test))
print("SVM classifier (trained with subset of training data): ")
print("%s of %s values correct." % (num_correct, len(y_test)))

SVM classifier (trained with subset of training data): 
8267 of 10000 values correct.


You should see the performance around 82 %. That is not bad, however, we can do better if we use all training data for building the model. This can take a while to finish.

In [39]:
svm_model.fit(X_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

Performance on the test data should be increased to around 94 %. 

In [40]:
# testing
predictions = [int(a) for a in svm_model.predict(X_test)]
num_correct = sum(int(a == y) for a, y in zip(predictions, y_test))
print("SVM classifier (trained with subset of training data): ")
print("%s of %s values correct." % (num_correct, len(y_test)))

SVM classifier (trained with subset of training data): 
9435 of 10000 values correct.


# 3. Converting SVM model to Core ML format

In [41]:
import coremltools

In [42]:
# we will use coremltools's converter, first we can check signature of the convert function: 
coremltools.converters.sklearn.convert?

In [43]:
coreml_svm = coremltools.converters.sklearn.convert(svm_model, 'hand_digit', 'predicted_digit')
# save the model
coreml_svm.save('digit_recognizer.mlmodel')

The coreml model should be visible in the directory where you have this notebook. Optionally, you can also add more information/metadata to your model.

In [16]:
# set model metadata
coreml_svm.author = 'Erkan Diken'
coreml_svm.license = 'BSD'
coreml_svm.short_description = 'Scikit-learn SVM classifier based handwritten digit recognizer '

# set feature descriptions manually
coreml_svm.input_description['hand_digit'] = 'handwritten digit'

# set the output descriptions
coreml_svm.output_description['predicted_digit'] = 'predicted digit'

# save the model
coreml_svm.save('digit_recognizer.mlmodel')

# 4. Load the Core ML model to make predictions 
Note: If you do not have MacOS version of 10.13. The predict method below will throw the following exception: Model prediction is only supported on macOS version 10.13.

Tested on MacOS High Sierra version 10.13.6

In [17]:
# load the model 
model = coremltools.models.MLModel('digit_recognizer.mlmodel')

We can check the the specs of your model.

In [18]:
model.get_spec

<bound method MLModel.get_spec of input {
  name: "hand_digit"
  shortDescription: "handwritten digit"
  type {
    multiArrayType {
      shape: 784
      dataType: DOUBLE
    }
  }
}
output {
  name: "predicted_digit"
  shortDescription: "predicted digit"
  type {
    int64Type {
    }
  }
}
predictedFeatureName: "predicted_digit"
metadata {
  shortDescription: "Scikit-learn SVM classifier based handwritten digit recognizer "
  author: "Erkan Diken"
  license: "BSD"
}
>

Documentation of the predict function points out that predict function accept inputs as dictionary object and key of the dictionary is the string that we have provided above. The same holds for the return value of the function.

In [19]:
model.predict?

In [20]:
digit_to_predict = dict()
digit_to_predict['hand_digit'] = X_test[0]

In [21]:
prediction = model.predict(digit_to_predict)

In [22]:
print (int(prediction['predicted_digit']))

7


The prediction output is the same with output of the scikit-learn SVM model for the first digit in test data. We can check the performance of coreml model on the whole test data.

In [27]:
print(X_test.shape[0])

10000


In [28]:
predictions = list()
for x in range(X_test.shape[0]):
    digit_to_predict['hand_digit'] = X_test[x]
    predictions.append(int(model.predict(digit_to_predict)['predicted_digit']))
print(len(predictions))

10000


In [29]:
num_correct = sum(int(a == y) for a, y in zip(predictions, y_test))
print("Core ml classifier: ")
print("%s of %s values correct." % (num_correct, len(y_test)))

Core ml classifier: 
9435 of 10000 values correct.


9435 of 10000 values correct. This is exactly the same as the SVM classifer before converted to coreml format.